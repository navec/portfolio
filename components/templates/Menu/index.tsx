import { HamburgerIcon, SmallCloseIcon } from "@chakra-ui/icons";
import { Button } from "@chakra-ui/react";
import { useState } from "react";

type MenuItem = { name: string };
type MenuProps = { items?: MenuItem[] };

const NavContent = ({
  closeMenu,
  items,
}: {
  closeMenu: () => void;
  items: MenuItem[];
}) => (
  <ul>
    <li>
      <Button variant="link" onClick={closeMenu}>
        <SmallCloseIcon />
      </Button>
    </li>
    {items.map(({ name }: { name: string }) => (
      <li key={name}>{name}</li>
    ))}
  </ul>
);

const NavIcon = ({ openMenu }: { openMenu: () => void }) => (
  <Button variant="link" onClick={openMenu}>
    <HamburgerIcon />
  </Button>
);

const defaultData = [
  { name: "Home" },
  { name: "Work" },
  { name: "About" },
  { name: "Contact" },
];

const Menu = ({ items = defaultData }: MenuProps) => {
  const [isOpen, setIsOpen] = useState<boolean>(false);

  const openMenu = () => setIsOpen(true);
  const closeMenu = () => setIsOpen(false);

  return (
    <>
      {isOpen && <NavContent closeMenu={closeMenu} items={items} />}
      {!isOpen && <NavIcon openMenu={openMenu} />}
    </>
  );
};

export default Menu;
