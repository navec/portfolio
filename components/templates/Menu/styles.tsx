import {
  Center,
  ChakraComponent,
  ComponentWithAs,
  HStack,
  Link,
  StackProps,
} from '@chakra-ui/react';
import styled from '@emotion/styled';

const Stack: ComponentWithAs<'div', StackProps> = ({
  children,
  ...props
}: StackProps) => {
  return (
    <HStack align="center" {...props}>
      {children}
    </HStack>
  );
};

export const Nav = styled(Stack)`
  background-color: ${({ theme }) => theme.colors.primary};
  color: ${({ theme }) => theme.colors.light};
`;

export const Li: ChakraComponent<'div', {}> = ({ children, ...props }) => (
  <Center flex={1} as="li" {...props}>
    <Link>{children}</Link>
  </Center>
);

export const NavItem = styled(Li)`
  height: ${({ theme }) => theme.space.rem[16]};
  font-size: ${({ theme }) => theme.fontSizes.lg};
`;
