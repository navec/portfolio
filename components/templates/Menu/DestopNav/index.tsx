import { MenuItemProps } from '@/components/templates/Menu';
import Link from 'next/link';

import * as S from './styles';

type DestopNavProps = MenuItemProps & { isSelected: boolean };

const DestopNav = ({ items }: { items: DestopNavProps[] }) => {
  return (
    <S.Layout>
      <S.Menu spacing={6}>
        {items.map(({ id, name, href, isSelected }) => (
          <S.MenuItem role="presentation" className={isSelected ? 'active' : undefined} key={id}>
            <Link href={href}>
              <a>{name}</a>
            </Link>
          </S.MenuItem>
        ))}
      </S.Menu>
    </S.Layout>
  );
};

export default DestopNav;
