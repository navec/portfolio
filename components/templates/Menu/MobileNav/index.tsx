import { useToggle } from '@/components/hooks/useToogle';
import { MenuItemProps } from '@/components/templates/Menu';
import Item from './MobileNavItem';

import * as S from './styles';

const YEAR = new Date().getFullYear();

type MobileNavProps = {
  items: (MenuItemProps & { isSelected: boolean })[];
  socials: MenuItemProps[];
};

const MobileNav = ({ items, socials }: MobileNavProps) => {
  const { isActive, toogle } = useToggle();

  return (
    <S.Layout>
      <S.MenuIcon className={isActive ? 'close' : 'show'} onClick={toogle} />

      <S.MenuContent
        direction="left"
        in={isActive}
        transition={{ exit: { duration: 0.4 } }}
      >
        <S.Label>Menu</S.Label>
        <S.Menu>
          {items.map(({ id, name, href, icon, isSelected }) => (
            <Item
              key={id}
              icon={icon}
              value={name}
              href={href}
              isSelected={isSelected}
            />
          ))}

          <S.Diviser />

          {socials.map(({ id, name, href, icon }) => (
            <Item key={id} icon={icon} value={name} href={href} />
          ))}
        </S.Menu>

        <S.Copyright>Â© {YEAR} - GAEL BATCHI</S.Copyright>
      </S.MenuContent>
    </S.Layout>
  );
};

export default MobileNav;
