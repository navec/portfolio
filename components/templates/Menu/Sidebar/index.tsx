import { MenuItemProps } from '@/components/templates/Menu';
import Link from 'next/link';

import * as S from './style';

type SidebarProps = MenuItemProps & { isSelected: boolean };

const Sidebar = ({ items }: { items: SidebarProps[] }) => {
  return (
    <S.Menu spacing={3}>
      {items.map(({ id, icon, href, isSelected }) => (
        <S.MenuItem className={isSelected ? 'active' : undefined} key={id}>
          <Link href={href}>
            <a>
              <S.Icon as={icon} />
            </a>
          </Link>
        </S.MenuItem>
      ))}
    </S.Menu>
  );
};

export default Sidebar;
