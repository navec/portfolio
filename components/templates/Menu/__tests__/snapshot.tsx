import { render } from "@testing-library/react";
import Menu from "..";

xit("renders menu with defaul menu", () => {
  // Given

  // When
  const { container } = render(<Menu />);

  // Then
  expect(container).toMatchSnapshot();
});
