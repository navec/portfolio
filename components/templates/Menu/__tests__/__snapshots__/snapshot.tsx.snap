// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders menu with defaul menu 1`] = `
<div>
  <ul>
    <li>
      Accueil
    </li>
    <li>
      Work
    </li>
    <li>
      About
    </li>
    <li>
      Contact
    </li>
  </ul>
</div>
`;
